src/integrations/unixIntegrationManager.ts

```diff
@@ -0,0 +1,158 @@
+      return !['steve', 'trivy'].includes(name);
+    });
+  }
+
+  // gets the names of the integrations that we want to symlink into the
+  // docker CLI plugin directory. They should all be of the form "docker-*".
+  async getDockerCliPluginNames(): Promise<string[]> {
+    return (await fs.promises.readdir(this.resourcesDir)).filter((name) => {
+      return name.startsWith('docker-');
+    });
+  }
+
+  protected async ensureIntegrationDir(desiredPresent: boolean): Promise<void> {
+    if (desiredPresent) {
+      await fs.promises.mkdir(this.integrationDir, { recursive: true, mode: 0o755 });
+    } else {
+      await fs.promises.rm(this.integrationDir, { force: true, recursive: true });
+    }
+  }
+
+  protected async ensureIntegrationSymlinks(desiredPresent: boolean): Promise<void> {
+    // get list of integrations in the resources directory
+    const integrationNames = await this.getIntegrationNames();
+
+    // create or remove the integrations
+    for (const name of integrationNames) {
+      const installationPath = path.join(this.resourcesDir, name);
+      const realizedPath = path.join(this.integrationDir, name);
+
+      await manageSymlink(installationPath, realizedPath, desiredPresent);
+    }
+  }
+
+  protected async ensureDockerCliSymlinks(desiredPresent: boolean): Promise<void> {
+    // ensure the docker plugin path exists
+    await fs.promises.mkdir(this.dockerCliPluginDir, { recursive: true, mode: 0o755 });
+
+    // get a list of docker plugins
+    const pluginNames = await this.getDockerCliPluginNames();
+
+    // create or remove the plugin links
+    for (const name of pluginNames) {
+      const integrationPath = path.join(this.integrationDir, name);
+      const dockerCliPluginPath = path.join(this.dockerCliPluginDir, name);
+
+      await manageSymlink(integrationPath, dockerCliPluginPath, desiredPresent, this.integrationDir);
+    }
+  }
+}
+
+// Ensures a symlink is either present or not present, while only changing it if
+// the target path of any existing symlink matches a search string. Idempotent.
+// @param srcPath The target path of the symlink.
+// @param dstPath The path of the symlink.
+// @param desiredPresent true to ensure the symlink is present; false to ensure it is not.
+// @param searchString The string that the existing symlink's target path must match
+//                     if changes are to be made to it. Default: resources/<platform>/bin
+export async function manageSymlink(srcPath: string, dstPath: string, desiredPresent: boolean, searchString?: string): Promise<void> {
+  let linkedTo: string;
+
+  searchString = searchString ?? path.join('resources', os.platform(), 'bin');
+
+  if (desiredPresent) {
+    try {
+      linkedTo = await fs.promises.readlink(dstPath);
+    } catch (error: any) {
+      if (error.code === 'ENOENT') {
+        await fs.promises.symlink(srcPath, dstPath);
+
+        return;
+      } else if (error.code === 'EINVAL') {
+        return;
```

Comment:
Does this error need to be surfaced to the user?

Reply:
This case occurs when the `dstPath` in question already exists, and is not a symlink. This is part of the normal usage of `manageSymlink` - the idea is that we make the symlink only if the user doesn't already have something there (either a normal file or a non-RD symlink). Of course, there is the potential for other errors, like insufficient permissions - in that case propagating an error is definitely appropriate. But this is intentional.

We could add a `console.debug` statement here though - that could be beneficial.